Tests for Trampoline Functions

cpetersmeier opened the pull request:
# Tests for Trampoline Functions
Trampoline functions allow overriding virtual functions of derived `C++` classes from within `Python`. This functionality should be tested to ensure consistent code quality. 

## Contents
- Modify and extend pybind11 wrappers
- Create tests for trampoline classes
- Cleanup `rostest_mtc`

cpetersmeier committed:
rostest_mtc: cleanup

    - Remove PyGenerator test.

cpetersmeier committed:
core wrapper: extend and rename

    - Rename trampoline virtual function arguments to (1) comply and (2)
      be identical to the argument names used in python tests.
    - Extend bindings through the inclusion of InterfaceState (used when
      deriving from 'Connecting') and constructors for the
      'Propagating-' classes

cpetersmeier committed:
rostest_trampoline: create tests for trampolines

    - Test parameters, class hierarchy, overriding virtual methods
      and insertion into task hierarchy.

rhaschke committed:
Fix wording

rhaschke committed:
kvargs -> kwargs (keyword args)

rhaschke committed:
prefer argument over global variables

rhaschke committed:
Avoid use of same identifier for different contexts

rhaschke reviewed (changes requested):
The ratio of actual test code to boilerplate code is rather bad. Actually, there are no automatic checks included yet, e.g. testing whether (or how often) a particular virtual method was called. Consider using [unittest's assertion methods](https://docs.python.org/3/library/unittest.html#assert-methods) instead of manually verifying print statements.

cpetersmeier committed:
rostest_trampoline: assert virtual method calls

    - mock.patch virtual methods to assert
      correct insertion into the task hierarchy

cpetersmeier commented:
Closing this as the relevant code has been moved to https://github.com/ros-planning/moveit_task_constructor/commits/wip-python-api.
